src/launchpad/size/insights/apple/image_optimization.py

```diff
@@ -0,0 +1,175 @@
+                    if res := self._check_heic_conversion(img, file_size):
+                        conversion_savings, heic_size = res.savings, res.optimized_size
+                elif fmt in {"heif", "heic"}:
+                    if res := self._check_heic_minification(img, file_size):
+                        minify_savings, minified_size = res.savings, res.optimized_size
+        except Exception as exc:
+            logger.error("Failed to process %s: %s", display_path, exc)
+            return None
+
+        if max(minify_savings, conversion_savings) < self.MIN_SAVINGS_THRESHOLD:
+            return None
+
+        return OptimizableImageFile(
+            file_info=source_object,
+            current_size=file_size,
+            minify_savings=minify_savings,
+            minified_size=minified_size,
+            conversion_savings=conversion_savings,
+            heic_size=heic_size,
+        )
+
+    def _check_minification(self, img: Image.Image, file_size: int, fmt: str) -> _OptimizationResult | None:
+        try:
+            with io.BytesIO() as buf:
+                save_params = {"optimize": True}
+                if fmt == "png":
+                    img.save(buf, format="PNG", **save_params)
+                else:
+                    work = img.convert("RGB") if img.mode in {"RGBA", "LA", "P"} else img
+                    work.save(buf, format="JPEG", quality=self.TARGET_JPEG_QUALITY, **save_params)
+                new_size = buf.tell()
+            return _OptimizationResult(file_size - new_size, new_size) if new_size < file_size else None
+        except Exception as exc:
+            logger.error("Minification check failed: %s", exc)
+            return None
+
+    def _check_heic_conversion(self, img: Image.Image, file_size: int) -> _OptimizationResult | None:
+        try:
+            with io.BytesIO() as buf:
+                img.save(buf, format="HEIF", quality=self.TARGET_HEIC_QUALITY)
+                new_size = buf.tell()
+            return _OptimizationResult(file_size - new_size, new_size) if new_size < file_size else None
+        except Exception as exc:
+            logger.error("HEIC conversion check failed: %s", exc)
+            return None
+
+    def _check_heic_minification(self, img: Image.Image, file_size: int) -> _OptimizationResult | None:
+        try:
+            with io.BytesIO() as buf:
+                img.save(buf, format="HEIF", quality=self.TARGET_HEIC_QUALITY)
+                new_size = buf.tell()
+            return _OptimizationResult(file_size - new_size, new_size) if new_size < file_size else None
+        except Exception as exc:
+            logger.error("HEIC minification check failed: %s", exc)
+            return None
+
+    def _iter_optimizable_files(self, files: Sequence[FileInfo]) -> Iterable[FileInfo]:
+        for fi in files:
+            if fi.file_type == "car":
+                yield from (c for c in fi.children if self._is_optimizable_image_file(c))
```

Comment:
oh, this is a cool syntax for going through the car file. I didn't know you could do that! very cool.

Reply:
Yea this is called a "generator" function since it yields values to iterators. In this case it does make the code a lot nicer to read since it handles the nested-ness of `car` files without having to modify the calling code.